nixos/modules/services/web-apps/dokuwiki.nix

```diff
@@ -0,0 +1,290 @@
+{ config, lib, pkgs, ... }:
+
+let
+
+  inherit (lib) mkEnableOption mkForce mkIf mkMerge mkOption recursiveUpdate types;
+
+  cfg = config.services.dokuwiki;
+
+  user = config.services.nginx.user;
+  group = config.services.nginx.group;
+
+  dokuwikiAclAuthConfig = pkgs.writeText "acl.auth.php" ''
+    # acl.auth.php
+    # <?php exit()?>
+    #
+    # Access Control Lists
+    #
+    ${toString cfg.acl}
+  '';
+
+  dokuwikiLocalConfig = pkgs.writeText "local.php" ''
+    <?php
+    $conf['savedir'] = '${cfg.stateDir}';
+    $conf['superuser'] = '${toString cfg.superuser}';
+    $conf['useacl'] = '${toString cfg.aclUse}';
+    ${toString cfg.extraConfig}
+  '';
+
+  dokuwikiPluginsLocalConfig = pkgs.writeText "plugins.local.php" ''
+    <?php
+    ${cfg.pluginsConfig}
+  '';
+
+  dokuwikiUsersAuthConfig = pkgs.writeText "users.auth.php" ''
+    # users.auth.php
+    # <?php exit()?>
+    #
+    # Userfile
+    #
+    # Format:
+    #
+    # login:passwordhash:Real Name:email:groups,comma,separated
+    ${toString cfg.users}
+  '';
+
+in
+{
+  options.services.dokuwiki = {
+    enable = mkEnableOption "DokuWiki web application.";
+
+    hostName = mkOption {
+      type = types.str;
+      default = "localhost";
+      description = "FQDN for the instance.";
+    };
+
+    stateDir = mkOption {
+      type = types.path;
+      default = "/var/lib/dokuwiki/data";
+      description = "Location of the dokuwiki state directory.";
+    };
+
+    acl = mkOption {
+      type = types.nullOr types.lines;
+      default = null;
+      example = "*               @ALL               8";
+      description = ''
+        Access Control Lists: see <link xlink:href="https://www.dokuwiki.org/acl"/>
+        Mutually exclusive with services.dokuwiki.aclFile
+      '';
+    };
+
+    aclFile = mkOption {
```

Comment:
I think this option should go. Or the `acl` option... but probably this one. Keeping the `acl` option encourages an immutable system, which is what we want. In the case that you can't use `acl` and need a file, presumably to manage things in a system mutable way, a note could be added to override `services.phpfpm.pools.dokuwiki.phpEnv.DOKUWIKI_ACL_AUTH_CONFIG` and you're good to go. Same for the `usersFile` option.

What do you think?

Reply:
I agree, but I would keep the ```usersFile``` as it contains passwords and you might want to keep it out the Nix store in most of the cases.

Reply:
@mmahut yes sorry I should have provided a more complete answer. We should omit the `users` option in favor of the `usersFile` option so we don't encourage people to store credentials in the nix store. There is always the fallback that users who absolutely insist on store credentials in the nix store can write `usersFile = pkgs.writeText "bad-idea.txt" '' credentials here '';`

How does this sound to you @1000101?

Reply:
@aanderse dropping users and keeping just usersFile so we don't have it in store sounds good so I'll definitely do that in a moment. But I'd opt for keeping the aclFile since it's really handy if you can edit it directly through the admin interface. Since we'd be keeping usersFile, there will already be a part which is mutable. I'd limit it to these two files only, though.

Reply:
@1000101 yeah, that is a good point... 
@mmahut or @Infinisil... any opinion on this?

Reply:
That is true, if we have password file mutable, the acl file can be mutable too. In that case, I vote for keeping acl as it is right now, but remove the ```users``` optiom.

Reply:
@mmahut I completely agree. @aanderse would you be ok with having it this way?

Reply:
Yeah that is fine with me. :+1:

Reply:
Great, I've incorporated and pushed these changes right now. Thanks a million to both of you for your wonderful suggestions and support!